scipy/_lib/_ccallback.py

```diff
@@ -0,0 +1,209 @@
+from . import _ccallback_c
+
+import types
+import ctypes
+
+PyCFuncPtr = ctypes.CFUNCTYPE(ctypes.c_void_p).__bases__[0]
+
+py_function_types = (types.MethodType, types.FunctionType)
+
+try:
+    import cffi
+    ffi = cffi.FFI()
```

Comment:
Beware that this can be costly:
```
>>> %time import cffi; cffi.FFI().CData
CPU times: user 24 ms, sys: 0 ns, total: 24 ms
Wall time: 21.7 ms
```